(TK-448) Add support for non-recursive directory watching

MosesMendoza opened the pull request:
* For the watcher protocol implementation, add a new field `recursive`. The field is initially populated as an atom with a nil value. Upon the first watch directory added, the field is set "permanently" to the value specified for the  :recursive option. We do so by enforcing that upon addition of any new directories to watch, the :recursive option must be the same as previously supplied, or we throw an exception. Also add a docstring to watch-new-directories! to clarify its behavior.

* Plumb recursive option through service. Update the required methods in the watch service to take the recursive setting into acount when deciding whether or not to recurse into directories that are watched and add them to the list of watched directories. With this, if a user specificies :recursive false, we limit watching to only the directory specified by `add-watch-dir!` and no others.

* Add new test utilities that allow us to check if a specific event - and only that event - have been registered by the watch service. We do this by immediately doing our comparison check on the first event that is registered.

  We are in the odd position of needing to test that something -doesn't happen-, not that it does. We can't just test that no events are registered, because that might happen for a number of reasons (ie the underlying watcher service failed). So instead we have a canary event alongside every operation that should not generate an event, and then we check that the only event that is registered is the canary event.

Commit:
(TK-448) add expected tests

They don't do anything yet but we know what we want to work

Signed-off-by: Moses Mendoza <moses@puppet.com>

Commit:
(TK-448) update options validation to expect boolean

We will now support both True and False for recursive, but require the option
be specified, so check that a boolean has been passed rather than asserting the
value matches true)

Signed-off-by: Moses Mendoza <moses@puppet.com>

Commit:
(TK-448) Add first test that exercises non-recursive functionality

As expected this fails.

Add new test utilities that allow us to check if a specific event - and only
that event - have been registered by the watch service. We do this by
immediately doing our comparison check on the first event that is registered.

We are in the odd position of needing to test that something -doesn't happen-,
not that it does. We can't just test that no events are registered, because
that might happen for a number of reasons (ie the underlying watcher service
failed). So instead we have a canary event alongside every operation that
should not generate an event, and then we check that the only event that is
registered is the canary event.

Signed-off-by: Moses Mendoza <moses@puppet.com>

Commit:
(TK-448) make recursive an attribute of the watcher

For the watcher protocol implementation, add a new field `recursive`. The field
is initially populated as an atom with a nil value. Upon the first watch
directory added, the field is set "permanently" to the value specified for the
:recursive option. We do so by enforcing that upon addition of any new
directories to watch, the :recursive option must be the same as previously
supplied, or we throw an exception. Also add a docstring to
watch-new-directories! to clarify its behavior.

Signed-off-by: Moses Mendoza <moses@puppet.com>

Commit:
(TK-448) Plumb recursive option through service

This commit updates the required methods in the watch service to take the
recursive setting into acount when deciding whether or not to recurse into
directories that are watched and add them to the list of watched directories.
With this, if a user specificies :recursive false, we limit watching to only
the directory specified by `add-watch-dir!` and no others.

This commit rounds out additional tests for this functionality.

Signed-off-by: Moses Mendoza <moses@puppet.com>

MosesMendoza commented:
Will squash commits on review/approval

MosesMendoza commented:
The test failing in travis passes for me locally... it doesn't appear that it _should_ be failing, so i'm wondering if there's some sort of timing thing going on

Commit:
(TK-448) test fix for create/modify

Its possible that on some implementations creating a file and writing to it are
separate? Try just 'touching' the file

Signed-off-by: Moses Mendoza <moses@puppet.com>

MosesMendoza commented:
Yep, now some unrelated test appears to be failing, and meanwhile:
```
☂ : lein -U test :all

lein test puppetlabs.trapperkeeper.services.watcher.filesystem-watch-service-test

Ran 10 tests containing 71 assertions.
0 failures, 0 errors.
```

puppetcla commented:
CLA signed by all contributors.

MosesMendoza commented:
~The test seems to consistently fail in travis so i'm going to close and try to repro that failure~ I haven't been able to reproduce locally, but I did re-run travis against the current master branch commit, c18b025, and the exact same test failed.

It appears to be registering both a create and modify entry for a file getting created in the `nested watch directories` test. Some conversation in TK-387 leads me to think this is a known issue?

from @justinstoller: 
> On Linux for every created file there is at least a CREATE and MODIFY event that occurs.

Though it wasn't clear what the outcome was.

MosesMendoza commented:
Oh now it passes.

justinstoller commented:
Had a conversation with Moses about this because I was somewhat unsure why we chose this route. Lemme see if I can put it down here and see if that's something folks think is fine going forward.

The current interface has a protocol function `add-watch-dir!` that takes an options map that is always expected to have the key `:recursive` with the value `true`. This implies that there will be a way to add individual directories (paths) to a watcher that could be recursive or not.

Now that we're implementing non-recursive watching we don't want to watch per path but per watcher, and the perceived cost of implementing per path recursive watching means we feel like it is prohibitively expensive to implement.

It seems like setting recursion per-watcher would be a natural fit for adding or modifying to the service protocol, `create-watcher` could default to calling `create-recursive-watcher` while the new consumer could call a `create-non-recursive-watcher` for example.

But, for backwards compatibility, we don't want to change the signature of `add-watch-dir!` and so we've created a config value that the watcher holds in an atom. That atom is set as a side effect the first time that `add-watch-dir!` is called so we don't have to put that config value in the protocol.

justinstoller commented:
So, it seems to me that an option on the `add-watch-dir!` function that sets per-watcher state should be moved to the interface that returns the watcher (despite being clojure this is still very Java-objecty, so I would think of the `create-watcher` protocol as the constructor). I don't think it makes *less* sense to move that option to a "constructor" or "setter" and have the current option to `add-watch-dir!` actually be optional (and deprecated). I think it should also be feasible to do this is a backwards compatible way.

However, I don't know if others agree with me on that point or if I'm missing something, if that's a bunch of work, or if we have the timeframe to do it in.

haus commented:
I prefer putting it in the constructor if that is doable in a non-gross backward compatible manner.

MosesMendoza commented:
I agree with the sentiment that the recursive option should be on the `create-watcher` "constructor". It only made sense to be on the `add-watch-dir!` function if you could actually specify the option on a per-watched directory basis.

That said, I think we're stuck between a rock and hard place here. I'm having a hard time coming up with something that accomplishes that with a good user experience that's backwards compatible.

We could add the recursive option to the constructor where it ideally would be, but what should happen when the recursive option is also passed in via `add-watch-dir!`?

I think its kind of messy. In order to be backwards compatible, the option passed to the function has to be authoritative. So if I've specified `:recursive true` to the constructor, and `:recursive false` to the `add-watch-dir!` function, we would disable recursive watching...which I think would be  surprising. Add a bit of extra complexity because we only override it if we somehow know that the option was originally specified as part of the constructor and not by a different call to `add-watch-dir!`, in which case we fail. We could say that since you've only ever been able to pass "false" as the recursive option, that's still the case - if you want "true" you have to specify it via the constructor. But then we have this vestigial options map alongside calls to `add-watch-dir!` floating around with `:recursive false` that doesn't do anything and is just ignored, which is also surprising UX.

I think the correct solution is to just stop supporting the recursive option on `add-watch-dir!` altogether, and only support it on the constructor function. It just doesn't make sense on that function. That's clearly a hard break we would do in a major release. In anticipate, we could add a warning or log entry on `add-watch-dir!` that the recursive option will be moving in an upcoming major release? Other than that though I'm hesitant to go further, and add more complexity/confusion with some mixture of options supported on both with a precedence behavior.

justinstoller commented:
Just to re-iterate, I don't what the perfect to be the enemy of the good, and if changing up the interface is too much work we can short circuit this convo. Having said that I disagree with...

> I think its kind of messy. In order to be backwards compatible, the option passed to the function has to be authoritative.

So, in current patch only the *first* call to `add-watch-dir!` is authoritative[1]. Subsequent calls require an option to be passed, but that option *must agree* with whatever option was passed to the first call of `add-watch-dir!`.  My thinking is that may be a more surprising behavior than if we could move that option to the "constructor" protocol function (but make it optional so older clients will continue to get a recursive watcher by default), and make the map to `add-watch-dir!` optional, but if it is supplied enforce that it must match the kind of watcher returned.


1 -  granted in our only consumer thus far we only call `add-watch-dir!` once, but I believe the interface is meant to support multiple invocations

MosesMendoza commented:
Chatted with @justinstoller - wasn't sure what the implications of:
> make the map to add-watch-dir! optional, but if it is supplied enforce that it must match the kind of watcher returned

That means we'll enforce effectively that a user can only specify a non-recursive watcher (value of `false`) via use of an option on the `create-watcher` function, the value of which will default to `true` (recursive). Then if the option is _also_ specified on `add-watch-dir!`, it also has to have a value of `false`. This isn't a backwards-incompatible change because users can upgrade cleanly - they won't have to change any option values of existing uses. And per @justinstoller's note the end result is likely an equivalent or lesser amount of surprising behavior.

I'm going to update the PR per this feedback.

Commit:
(TK-448) add recursive option to create-watcher

* Per feedback on
  https://github.com/puppetlabs/trapperkeeper-filesystem-watcher/pull/17,
  update to support specifying the recursive option on the `create-watcher`
  function. It is not intuitive to support this option on the `add-watch-dir!`
  function when the option cannot actually vary on a per-directory basis.

* If no options are specified to `create-watcher`, the recursive value
  defaults to `true`. Prior to the work for TK-448, the recursive option was
  required on `add-watch-dir!`, and only with a value of true. Therefore on
  upgrade, this commit should result in no behavior change - any calls to
  `add-watch-dir!` should be accompanied by a {:recursive true} option aligned
  with the default for `create-watcher`. To specify non-recursive watching,
  {:recursive false} must be supplied to the `create-watcher` function (and no
  option specified or false specified to `add-watch-dir!`).

* While the `add-watch-dir!` function still accepts options
  containing a recursive value, this is only used to validate that the value is
  identical to the one already specified to the `create-watcher` function. We
  do so to avoid any possible confusion because we do not support different
  values for this option on different directories.

* To retain full backwards-compatibility, multiple arities are added to several
  functions.

Signed-off-by: Moses Mendoza <moses@puppet.com>

MosesMendoza commented on src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj:
```diff
@@ -53,31 +53,53 @@
+
+  (add-watch-dir!
     [this dir options]
     (validate-watch-options! options)
```
should this have a deprecation warning emitted, ie `log/debug "passing options ..."`?

justinstoller replied:
yeah

MosesMendoza replied:
Do you know if we will always automatically emit that deprecation because of our use for HA? @haus  can we update HA to use the new form?

MosesMendoza replied:
debug message added

MosesMendoza commented on src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj:
```diff
@@ -53,31 +53,53 @@
+        (DirWatchUtils/registerRecursive watch-service [watched-path])
+        (DirWatchUtils/register watch-service watched-path))))
+
+  (add-watch-dir!
```
Looked odd to me at first but apparently multi-arity protocol function definitions are achieved via separate discrete definitions.

MosesMendoza commented on src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj:
```diff
@@ -53,31 +53,53 @@
-    {:watch-service (.newWatchService (FileSystems/getDefault))
-     :callbacks (atom [])}))
+  ([]
+   (create-watcher {:recursive true}))
```
What's the standard for spacing/indentation in clojure?

justinstoller replied:
Two spaces unless it's a list with a single item on the first line, then one.

We follow this https://github.com/bbatsov/clojure-style-guide
relatively closely, with three noticeable exceptions being we don't usually make anything private, prefer schema to pre/post assertions, and always use lists of `<package><space><one or more classes>` in imports (rather than a vectors of `<package><dot><class>`).

MosesMendoza replied:
ok thanks

justinstoller commented:
I think this looks good, though I haven't built these changes out into a server build (with the other complementary PRs) to see if it solves the problem it was intended to solve.

It'd be great to get a review/merge from @haus or someone else who is, or wants to be, familar with the code base.

haus commented on test/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_service_test.clj:
```diff
@@ -43,6 +49,25 @@
+  event. If so, returns the expected event. Otherwise returns contents of dest."
+  [dest expected-event]
+  (let [start-time (System/currentTimeMillis)]
+    (loop []
```
It looks like this could be a `while` instead of a `loop`/`recur` as there are no bindings for the loop.

MosesMendoza replied:
So I could do this but `elapsed-time` of the recur condition `(< elapsed-time wait-time)` is dynamically redefined/rebound every iteration which appears harder with `while` and we're also waiting for the atom value to be populated.
I could do: `(while (< (- (System/currentTimeMillis) start-time) wait-time)` and avoiding binding `elapsed-time` but that seems less readable to me. It basically ends up being something like "just wait until one of these conditions is true", ie:
```
  (let [start-time (System/currentTimeMillis)]
    (while (and (empty? @dest)
                (< (- (System/currentTimeMillis) start-time) wait-time)))
    (if (empty? @dest)
      (do
        (println "timed-out waiting for events")
         @dest)
      (if (exactly-matches-event? dest expected-event)
        expected-event
        @dest)
```
this seems really weird to me so i'm sure there's a better way

MosesMendoza replied:
chatted with @haus in meatspace, need to use binding in loop

MosesMendoza replied:
updated

haus commented on src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj:
```diff
@@ -53,31 +53,53 @@
-     :callbacks (atom [])}))
+  ([]
+   (create-watcher {:recursive true}))
+  ([options]
```
You might consider using destructuring here in the fn args.

MosesMendoza replied:
why?

haus replied:
It makes it clear which options the function is caring about.

It would matter more if there were a bunch of keys being pulled out of the map, in this case either way is fine.

MosesMendoza replied:
ah sure. I wasn't sure but that seems reasonable

MosesMendoza replied:
@haus how do I destructure this in order to retain both a reference to `options` since we use it as well as the value of `:recursive`?

MosesMendoza replied:
per comment from @justinstoller this is  `{:keys [recursive] :as options}`

MosesMendoza replied:
updated

Commit:
(TK-448) whitespace fixes

* Adjust whitespace / indentation / forms per styleguide

Signed-off-by: Moses Mendoza <moses@puppet.com>

Commit:
(TK-448) destructure options to create-watcher

Signed-off-by: Moses Mendoza <moses@puppet.com>

Commit:
(TK-448) debug message for using options on add-watch-dir!

Signed-off-by: Moses Mendoza <moses@puppet.com>

Commit:
(TK-448) use binding in loop in test function

per feedback / idiomatic

Signed-off-by: Moses Mendoza <moses@puppet.com>

Commit:
(TK-448) Changelog for 1.1.0

* Add a changelog entry reflecting the content for the 1.1.0 feature release.

Signed-off-by: Moses Mendoza <moses@puppet.com>

haus commented:
@KevinCorcoran 💩

haus commented on src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj:
```diff
@@ -53,31 +53,55 @@
+   (map->WatcherImpl
+     {:watch-service (.newWatchService (FileSystems/getDefault))
+      :callbacks (atom [])
+      :recursive (atom recursive)})))
```
Doesn't matter too much, but this doesn't need to be an `atom` does it? It will never change.

MosesMendoza replied:
Yeah I think the atom bit is leftover from when we were first spiking this out

haus replied:
right

haus commented:
oh travis and the infinite sadness.

haus commented:
I had one question but it doesn't need to block merging this.

MosesMendoza commented:
yeah thats the same test that passes locally but that intermittently fails or passes depending on something in travis
